README.md
lua-ruuvi
=========

Lua bindings for accessing data from the RuuviTracker via the client API

## Contents

1. Licence
2. About
3. Requirements
4. Installation
5. Usage


# 1 Licence

lua-ruuvi is free software: you can redistribute it and/or modify it under 
the terms of the GNU General Public License as published by the Free 
Software Foundation, either version 3 of the License, or (at your option) 
any later version.

lua-ruuvi is distributed in the hope that it will be useful, but WITHOUT ANY 
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS 
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more 
details.

You should have received a copy of the GNU General Public License along with 
lua-ruuvi. If not, see [www.gnu.org][1].

See COPYING for the full licence.

  [1]: http://www.gnu.org/licenses/


# 2 About

This project aims to provide a complete set of Lua bindings to access the 
various data provided by the [RuuviTracker][] server via its [client API][api].

  [RuuviTracker]:  http://www.ruuvitracker.fi/              "RuuviTracker's website"
  [api]:           http://wiki.ruuvitracker.fi/wiki/UI-API  "UI-API on the RuuviTracker wiki"


# 3 Requirements

lua-ruuvi is built for Lua 5.2, but should presumably work with 5.1 without 
any changes.

Required libraries and other things:

 * [dkjson](http://chiselapp.com/user/dhkolf/repository/dkjson/home "dkjson's homepage")
 * [lua-curl](http://msva.github.com/lua-curl/ "lua-curl on Github")


# 4 Installation

As of currently there has been made no package of the library. This will not 
be prioritised until it reaches some sort of usability.


# 5 Usage

See `testlib.lua` for examples on how the library is tested (and thus, how it is used).

TBW.

//EOF


ChangeLog
Changelog for lua-ruuvi

 + 2013-01-11
 
  Initialised repo

 + 2013-01-12
 
  Began structuring the library and figuring out how Lua 5.2 works

 + 2013-01-14
 
  Finished debugging ping for now; it works. Brought tracker and event listing
  functions into working condition (needs refactoring).


events.lua
#!/usr/bin/env lua
-- events.lua - Functions relating to events
-- Copyright Stæld Lakorv, 2013 <staeld@illumine.ch>
-- This file is part of lua-ruuvi.
-- lua-ruuvi is released under the GPLv3 - see COPYING
local M = {}

-- TODO: Refactor these into one/several smaller functions; reuse code

-- events.getList(): get list of events, optionally only events specified by ids
function M.getList(servobj, idString)
    local path = "events"
    if idString then path = path .."/".. idString end
    local c = servobj._c
    c:setopt_url(servobj.url .. path)
    local f = io.tmpfile()
    c:perform({ writefunction = function(str) f:write(str) end })
    local obj = aux.readjson(f)
    f:close()
    return obj.events
end

-- events.getFor(): retrieve events related to one or more trackers
function M.getFor(servobj, idString, paramArray)
    local path = "trackers/" .. idString .. "/events"
    if paramArray then path = aux.url_addParams(path, paramArray) end
    local c = servobj._c
    c:setopt_url(servobj.url .. path)
    local f = io.tmpfile()
    c:perform({ writefunction = function(str) f:write(str) end })
    local obj = aux.readjson(f)
    f:close()
    return obj.events
end

-- events.getLatestFor(): retrieve the latest events related to one or more trackers
function M.getLatestFor(servobj, idString, paramArray)
    local path = "trackers/" .. idString .. "/events/latest"
    if paramArray then path = aux.url_addParams(path, paramArray) end
    local c = servobj._c
    c:setopt_url(servobj.url .. path)
    local f = io.tmpfile()
    c:perform({ writefunction = function(str) f:write(str) end })
    local obj = aux.readjson(f)
    f:close()
    return obj.events
end

return M
-- EOF


helpers.lua
#!/usr/bin/env lua
-- helpers.lua - Auxilliary functions for lua-ruuvi
-- Copyright Stæld Lakorv, 2013 <staeld@illumine.ch>
-- This file is part of lua-ruuvi.
-- lua-ruuvi is released under the GPLv3 - see COPYING

-- Prerequisites
local curl = require("cURL")

-- Declare module container
local aux = {}

function aux.throwError(err)
    error(err)
    --io.stderr:write("error: ", err, "\n")
    --os.exit(1)
end

-- aux.connObj(): create a new curl connection object with library useragent
function aux.connObj(url, path)
    path = path or ""
    local c = curl.easy_init()
    c:setopt_useragent(libName .."/".. libVersion)
    c:setopt_url(url .. path)
    return c
end

-- Ping server via the API
function aux.ping(servobj)
    local pingTime = os.time(os.date("!*t"))    -- Ensure it's in UTC

    local c = servobj._c        -- Clone the server's connection object
    c:setopt_url(servobj.url .. "ping")

    local f = io.tmpfile()      -- Write api output to tmpfile
    c:perform({ writefunction = function(str) f:write(str) end })
    local recv = os.time(os.date("!*t"))
    local obj  = aux.readjson(f)
    f:close()
    if obj and obj.time then
        -- Calculate ping and time difference
        local serverTime = os.time(aux.toTimeTable(obj.time))
        local servdiff   = os.difftime(recv, serverTime)
        local lag        = os.difftime(recv, pingTime)
        print("DEBUG Ping successful! Turnabout lag: " .. lag .. ", time difference: " .. servdiff)
        return true, lag, servdiff
    else
        print("DEBUG Ping did not output proper response!")
    end
end

-- aux.url_addParams(): add parametres to the API path
function aux.url_addParams(path, paramArray)
    local paramsAdded = false
    for param, value in pairs(paramArray) do
        if not paramsAdded then
            path = path .."?".. param .."=".. value
            paramsAdded = true
        else
            path = path .."&".. param .."=".. value
        end
    end
    return path
end

-- aux.readjson(): reads and parses json from a temporary file
function aux.readjson(filehandle)
    filehandle:seek("set")
    local str = filehandle:read("*a")
    local obj, pos, err = json.decode(str)
    if err then aux.throwError(err)
    else return obj end
end

-- aux.toTimeTable(): creates a standard time table from a string
function aux.toTimeTable(str)
    local t = {}
    t.year, t.month, t.day = str:match("(%d%d%d%d)%-(%d%d)%-(%d%d)")
    t.hour, t.min, t.sec, t.off, t.offh, t.offm = str:match("T(%d%d):(%d%d):(%d%d).-([+-])(%d%d)(%d%d)")
    -- Compensate for timezones, return time in UTC
    if t.off == "+" then
        t.hour = t.hour + t.offh
        t.min  = t.min  + t.offm
    elseif t.off == "-" then
        t.hour = t.hour - t.offh
        t.min  = t.min  - t.offm
    end
    return t
end

-- aux.toTimeString(): creates a standard time string from a table
function aux.toTimeString(t)
    os.date("!%FT%T%z", os.time(t))
end

return aux
-- EOF


servobj.lua
#!/usr/bin/env lua
-- servobj.lua - Server object creation for lua-ruuvi
-- Copyright Stæld Lakorv, 2013 <staeld@illumine.ch>
-- This file is part of lua-ruuvi.
-- lua-ruuvi is released under the GPLv3 - see COPYING
local M = {}

function M.new(url)
    if type(url) ~= "string" then aux.throwError("not a string") end
    local obj = {}
    if not url:find(vars.apiPath:gsub("([%.%-%+])", "%%%1")) then
        url = url:gsub("/$", "") .. vars.apiPath
    end

    obj.url = url
    obj._c  = aux.connObj(url)

    obj._c:setopt_useragent(libName .."/".. libVersion)

    function obj:ping()
        return aux.ping(self)
    end

    -- Events
    function obj:events(idString, paramArray)
        return events.getList(self, idString, paramArray)
    end
    -- Trackers
    function obj:trackerId(name)
        return trackers.getId(self, name)
    end
    function obj:trackerCode(id)
        return trackers.getCode(self, id)
    end
    function obj:trackerName(id)
        return trackers.getName(self, id)
    end
    function obj:trackers(idString, paramArray)
        return trackers.getList(self, idString, paramArray)
    end
    function obj:eventsFor(idString, paramArray)
        return events.getFor(self, idString, paramArray)
    end
    function obj:latestFor(idString, paramArray)
        return events.getLatestFor(self, idString, paramArray)
    end
    
    return obj
end

return M
-- EOF


testlib.lua
#!/usr/bin/env lua
-- testlib.lua - A file to load and test the lua-ruuvi library
-- Copyright Stæld Lakorv, 2013 <staeld@illumine.ch>
-- This file is part of lua-ruuvi.
-- lua-ruuvi is released under the GPLv3 - see COPYING

ruuvi = require("ruuvi")

url = "http://dev-server.ruuvitracker.fi/api/v1-dev/" -- Testing server
server = ruuvi.new(url)

print("Testing direct call to ruuvi.aux.ping()")
--ruuvi.aux.ping(url)

print("Testing server object creation and pinging")
server:ping()

print("Testing tracker listing")
for i, t in ipairs(server:trackers()) do
    for k, v in pairs(t) do print(k,v) end
end

print("Testing event listing")
local arr = server:events()
for i, t in ipairs(arr) do
    for k, v in pairs(t) do print(k,v) end
end

print("Testing fetching of tracker IDs")
print(server:trackerId("fooo"), server:trackerId("sepeto"))
print(server:trackerName("1"), server:trackerCode("4"))

print()
print("Last activity reported by tracker 'foobar'")
id = server:trackerId("foobar")
timestring = server:trackers(id)[1].latest_activity
t = ruuvi.aux.toTimeTable(timestring)

trackerTime = os.time(t)
now = os.time(os.date("!*t"))
diff = os.difftime(now, trackerTime)

days = diff / ( 60*60 * 24 )
if days < 1 then
    days = days * 24
    days = string.format("%.1f hours", days)
else
    days = string.format("%.1f days", days)
end
print(timestring, ", which is " .. days .. " ago.")

-- EOF


trackers.lua
#!/usr/bin/env lua
-- trackers.lua - Tracker-related functions for lua-ruuvi
-- Copyright Stæld Lakorv, 2013 <staeld@illumine.ch>
-- This file is part of lua-ruuvi.
-- lua-ruuvi is released under the GPLv3 - see COPYING
local M = {}

-- trackers.getList(): get trackers, optionally narrowed by ids
function M.getList(servobj, idString, paramArray)
    local path = "trackers"
    if idString then
        path = path .."/".. idString
    end
    if paramArray then path = aux.url_addParams(path, paramArray) end
    -- local c = aux.connObj(server, path)
    local c = servobj._c
    c:setopt_url(servobj.url .. path)
    local f = io.tmpfile()
    c:perform({ writefunction = function(str) f:write(str) end })
    local obj = aux.readjson(f)
    f:close()
    return obj.trackers
end

-- TODO: Refactor/fix these

-- trackers.getId(): get the id of a tracker with given code/name
function M.getId(servobj, name, paramArray)
    local list = M.getList(servobj, nil, paramArray)
    for i, t in ipairs(list) do
        if t.tracker_code == name then return t.id end
    end
    -- Try again, with real names instead of id-strings
    for i, t in ipairs(list) do
        if t.name == name then return t.id end
    end
    return nil
end
-- trackers.getCode(): return tracker_code for a single id
function M.getCode(servobj, id, paramArray)
    local list = M.getList(servobj, nil, paramArray)
    return M.handleGet(list, "tracker_code", id)
end
-- trackers.getName(): return tracker's name for a single id
function M.getName(servobj, id, paramArray)
    local list = M.getList(servobj, nil, paramArray)
    return M.handleGet(list, "name", id)
end

-- trackers.handleGet(): auxilliary function to handle lists
function M.handleGet(list, item, id)
    for i, t in ipairs(list) do
        if t.id == id then return t[item] end
    end

end

return M
-- EOF


ruuvi.lua
#!/usr/bin/env lua
-- ruuvi.lua - Bindings for accessing data from the RuuviTracker via the client API
-- Copyright Stæld Lakorv, 2013 <staeld@illumine.ch>
--
-- This file is part of lua-ruuvi.
--
-- lua-ruuvi is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.
--
-- lua-ruuvi is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with lua-ruuvi. If not, see <http://www.gnu.org/licenses/>.

libName, libVersion = "lua-ruuvi", "0.0.1"

if _VERSION == "Lua 5.2" then
    -- Fix the problem that our libraries may be in a 5.1 directory
    package.path = package.path ..";".. package.path:gsub("5%.2", "5.1")
end

-- Module wrapper
local M = {}

-- Prerequisites
json        = require("dkjson")
curl        = require("cURL")
vars        = require("vars")
aux         = require("helpers")  -- Load these first; used other places
events      = require("events")
trackers    = require("trackers")
local sobj  = require("servobj")

-- TODO: Figure out a sane way to return the module properly while preserving access from within the modules themselves when loading above
return { vars = vars, aux = aux, events = events, trackers = trackers, new = sobj.new }    -- Provide the module to the caller

-- EOF
